syntax = "proto3";
package heron.proto.system;

option java_package = "com.twitter.heron.proto.system";
option java_outer_classname = "PhysicalPlans";

import "topology.proto";

// This file defines the physical plan of a topology.
// It reflects the dynamic running state of a topology.

message StMgr {
  string id = 1;
  string host_name = 2;
  // This is the data port for communication
  // accross the containers(i.e. peer stmgrs)
  int32  data_port = 3;
  // This is the port for communication
  // within the container(i.e. local instances)
  int32  local_data_port = 8;
  string local_endpoint = 4;
  string cwd = 5; // current working directory
  int32 pid = 6; // PID of process

  // Http port to connect to heron-shell.
  // The port is not used by stmgr itself, except to
  // pass it on to tmaster to be included in pplan.
  int32 shell_port = 7;
}

message InstanceInfo {
  int32 task_id = 1; // global
  int32 component_index = 2; // specific to this component
  string component_name = 3;
  repeated string params = 4;
}

message Instance {
  string instance_id = 1;
  string stmgr_id = 2;
  InstanceInfo info = 3;
}

message PhysicalPlan {
  // Note that heron.proto.api.Topology, included in PhysicalPlan,
  // reflect the actual dynamic running state of the topology.
  // Query this value guarantees the correct running state of the topology.
  heron.proto.api.Topology topology = 1;
  repeated StMgr stmgrs = 2;
  repeated Instance instances = 3;
}
